src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -142,28 +154,70 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
                     continue;
                 }
                 if (version != -1) {
+                    // We found one that can be allocated!
                     numberOfAllocationsFound++;
-                    if (highestVersion == -1) {
-                        nodesWithHighestVersion.add(node);
+                    if (version > highestVersion) {
                         highestVersion = version;
+                    }
+                    // If we're allowed to recover this on any node, always
+                    // treat the node as a candidate
+                    if (recoverOnAnyNode(idxSettings)) {
+                        nodesWithVersion.put(node, version);
                     } else {
+                        // If we've found a new "best" candidate, clear the
+                        // current candidates and add it
                         if (version > highestVersion) {
-                            nodesWithHighestVersion.clear();
-                            nodesWithHighestVersion.add(node);
-                            highestVersion = version;
+                            nodesWithVersion.clear();
+                            nodesWithVersion.put(node, version);
                         } else if (version == highestVersion) {
-                            nodesWithHighestVersion.add(node);
+                            // If the candidate is the same, add it to the
+                            // list, but keep the current candidate
+                            nodesWithVersion.put(node, version);
                         }
                     }
                 }
             }
+            // Now that we have a map of nodes to versions along with the
+            // number of allocations found (and not ignored), we need to sort
+            // it so the node with the highest version is at the beginning
+            List<DiscoveryNode> nodesWithHighestVersion = Lists.newArrayList();
+            nodesWithHighestVersion.addAll(nodesWithVersion.keySet());
+            CollectionUtil.timSort(nodesWithHighestVersion, new Comparator<DiscoveryNode>() {
+                @Override
+                public int compare(DiscoveryNode o1, DiscoveryNode o2) {
+                    if (nodesWithVersion.get(o1) < nodesWithVersion.get(o2)) {
+                        return 1;
+                    } else if (nodesWithVersion.get(o1) == nodesWithVersion.get(o2)) {
+                        return 0;
+                    } else {
+                        return -1;
+                    }
+                }
+            });
+
+            if (logger.isDebugEnabled()) {
+                logger.debug("[{}][{}] found {} allocations of {}, highest version: [{}]",
+                        shard.index(), shard.id(), numberOfAllocationsFound, shard, highestVersion);
+            }
+            if (logger.isTraceEnabled()) {
```

Comment:
No, I want both logging messages to be written if TRACE logging is on, and if I use `else if` the second log message (the candidates) is skipped because debug is enabled.